src/chain/candidate.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::OnlinePayload;
+use crate::{id::PublicId, utils::LogIdent, utils::XorTargetInterval};
+use log::LogLevel;
+use std::collections::BTreeSet;
+
+/// A candidate (if any) may be in different stages of the resource proof process.
+/// When we consensus to accept them for resource proof, move to `AcceptedForResourceProof`
+/// with the value all elder expect to get from `CandidateInfo`.
+/// If we consensus to refuse them: reset to None.
+/// If we consensus to accept them: move to ApprovedWaitingSectionInfo until we are available to
+/// accept a new candidate, at which point reset to None.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone)]
+pub enum Candidate {
+    /// No-one is currently in the resource proof process.
+    /// We can take on a new candidate on consensus of an `ExpectCandidate` event.
+    None,
+    /// We accepted a candidate to perform resource proof in this section. We are waiting for
+    /// them to send their `CandidateInfo` before starting the actual resource proof.
+    AcceptedForResourceProof {
+        target_interval: XorTargetInterval,
+        old_public_id: PublicId,
+    },
+    /// We consensused the candidate online. We are waiting for the SectionInfo to consensus
+    /// and this new node to start handling events before allowing a new candidate.
+    ApprovedWaitingSectionInfo { new_pub_id: PublicId },
+}
+
+impl Candidate {
+    /// Return true if already has a candidate
+    pub fn has_resource_proof_candidate(&self) -> bool {
```

Comment:
Because all of these public function are forwarded from chain, I'd rather keep the candidate in them. Last time we had similar discussion we decided that it was better to have the forwarding function and the function have the same name, so I apply the same logic (not 100% sure it is the best, or if we should apply the same logic).

Reply:
- Repeating the name of the receiver in the method name is unidiomatic (see the std library: there is `Vec::len`, not `Vec::vec_len`, `File::open`, not `File::open_file`, `slice::sort_by_key` not `slice::sort_slice_by_key`, etc... 
- I don't think the fact that these function are only forwarded to matter. In fact, this type should not even be aware of how it's used since that violates separation of concerns. 
- Also this sets a weird precedent. Should we now follow this pattern in the future in similar situations? If so when? Always? Sometimes? etc...
- Ultimately it just makes the code verbose for no good reason. Verbosity is good if it helps clarity, but that is dubious in this case.

Anyway, up to you. If you insist on this, I won't block the merge over it.

Reply:
I'll fix it up, it make sense to me. Probably interpreted some previous review wrong.

Reply:
Here are the fixed name in candidate:
The forwarding function will still have candidate in the name but some will likely be updated too.

Does that seem ok?

```
    /// Return true if no candidate info
    pub fn is_none(&self) -> bool {

    /// Forget about the current candidate.
    pub fn reset(&mut self) {

    /// Forget about the current candidate if it is a member of the given section.
    pub fn reset_if_member_of(&mut self, members: &BTreeSet<PublicId>) {

    /// Our section decided that the candidate should be resource proofed next.
    /// Pre-condition: is_none.
    pub fn accept_for_resource_proof(
        &mut self,
        old_public_id: PublicId,
        target_interval: XorTargetInterval,
    ) {


    /// Try to accept as memeber.
    /// If the candidate was already purged or is unexpected, return false.
    /// Otherwise marks the candidate as `ApprovedWaitingSectionInfo`.
    pub fn try_accept_as_member(&mut self, online_payload: &OnlinePayload) -> bool {

    /// Return the target interval if we are resource proofing for that old PublicId.
    pub fn matching_target_interval(&self, old_public_id: &PublicId) -> Option<&XorTargetInterval> {


    /// The public id of the candidate we are resource proofing.
    pub fn old_public_id(&self) -> Option<&PublicId> {


    /// The target interval of the candidate we are resource proofing.
    fn target_interval(&self) -> Option<&XorTargetInterval> {

    /// Logs info about ongoing candidate state, if any.
    pub fn show_status(&self, log_ident: &LogIdent) {
```

Reply:
Perfect!